### PR TITLE
[NNX] NNX migration (11/N): set pure_nnx / enable_nnx / pure_nnx_decoder defaults to True

### PR DESCRIPTION
**Pull request by ecnal-cienet**
## NNX Migration Route Map
1. ✅ Add NNX scaffolding: `pure_nnx` flag, `init_state_fn`, `TrainStateNNX`, NNX utils. Linen workflow unchanged. (PR #3427)
2. ✅ NNX sharding utilities: `get_abstract_state_nnx`, `get_named_sharding_nnx`, `set_named_sharding_nnx`, `get_partition_spec_nnx`, `get_mesh_from_config`. (PR #3470)
3. ✅ NNX fully supported end-to-end: `TrainStateNNX`, model creation, gradient accumulation, checkpointing, and training loop dispatch. (PR #3500)
4. ✅ Sharding diagnostics on NNX + post-training bugfixes that surfaced once the NNX path was exercised end-to-end. (PR #3652)
5. ✅ NNX correctness fixes, feature enablements, vocab tiling on NNX.
6. ✅ NNX-native DPO.
7. ✅ NNX-native MaxEngine inference. (PR #3821)
8. ✅ NNX-native LoRA + GRPO. (PR #3824)
9. ✅ NNX-aware QK-Clip + remaining checkpoint utilities. (PR #3836)
9.5. ✅ NNX + AQT in MaxEngine + serve-mode reload + gpt3 prefill fix. (PR #3844)
10. ✅ Vocab tiling `custom_vjp` for NNX. (PR #3849)
11. 🔄 **[This PR]** Flip `enable_nnx`, `pure_nnx`, `pure_nnx_decoder` from `False` → `True` in `base.yml`. Bundle the NNX-only fixes that surface once `pure_nnx=True`.
12. ❌ Delete Linen-specific code paths and NNX compatibility flags. (PR #4038)

## Description

PR6–PR10 promoted every routed-to-Linen feature to NNX-native; #2885 added NNX-native pipeline parallelism; #4040 added Qwix on NNX. This PR flips the three defaults in `base.yml` and bundles the NNX-only fixes that surface once `pure_nnx=True`.

## Changes

**`src/maxtext/configs/base.yml`** — flip defaults
- `enable_nnx: False → True`, `pure_nnx: False → True`, `pure_nnx_decoder: False → True`.

**`src/maxtext/utils/sharding.py`** — Zero-1 on flat `nnx.State`
- New `build_zero1_input_state_mesh_shardings` overlays `Param`-leaf shardings on the flat `nnx.State`. The Linen call `state_mesh_shardings.replace(params=...)` only works on `TrainState`.

**`src/maxtext/trainers/pre_train/train.py`, `train_compile.py`** — NNX dispatch
- AOT compile dispatches to the Zero-1 NNX builder when `pure_nnx=True`.
- Pops `Intermediate` sown variables before grad so MTP auxiliary losses aren't differentiated as part of the main loss.

**`src/maxtext/trainers/diloco/diloco.py`, `src/maxtext/common/checkpointing.py`** — DiLoCo under NNX
- `DiLoCoTrainState.merge` / `.split` use `nnx.split`, guarded against double-merging.
- `maybe_save_checkpoint` reads `state.step` under `enable_diloco`, otherwise `state.optimizer.step`.
- `replace_nnx_model_params` identifies "model" leaves by path via `tree_flatten_with_path`. Preserves the original `treedef` so `lax.cond` branches still match, and is robust to future key additions to `inner_state` (addresses [#3526 review comment](https://github.com/AI-Hypercomputer/maxtext/pull/3526#discussion_r3405987983)).

**`src/maxtext/utils/generate_param_only_checkpoint.py`** — NNX param-only restore
- Pure-dict restore (`{"value": ...}` wrapping), opt_state path skipping, bf16 cast skipping rng leaves.

**`src/maxtext/inference/maxengine/maxengine.py`** — drop Linen-vs-NNX parity asserts
- NNX-only prefill/decode/cache assertions stay.

**`src/maxtext/layers/nnx_wrappers.py`** — modernize `ToLinen.__call__`
- Drop the `_refresh_variable_trace_state` private-state workaround; use idiomatic `nnx.split` / `nnx.update` / `nnx.merge` and filter unknown paths before assignment (addresses [#3526 review comment](https://github.com/AI-Hypercomputer/maxtext/pull/3526#discussion_r3280456243)).

**`src/maxtext/layers/quantizations.py`** — OSS qwix import fix from #4040
- `from qwix._src.utils import flax_util` → `from qwix._src import flax_util`. PR#4040 (Copybara) referenced the Google-internal `qwix._src.utils` path; OSS qwix has `flax_util` directly under `_src/`. Unblocks OSS CI.

**`src/maxtext/utils/{muon_utils,qk_clip_utils,train_utils}.py`** — NNX-shape adjustments
- `muon_utils.get_muon_weight_dimension_numbers` dispatches by NNX-vs-Linen state shape.
- `qk_clip_utils` broadcasts over the correct axis under NNX.
- `train_utils.jit_train_step` threads `dropout_rng=None` on the NNX path.

**`src/maxtext/trainers/post_train/sft/train_sft_native.py`** — SFT NNX path
- `nnx.split(state)` before `jit`, dropout-rng threaded conditionally (mirrors the pre-train path).

**`src/maxtext/checkpoint_conversion/{to_maxtext.py, utils/utils.py}`** — shared helper
- Route path-key parsing through the shared `param_key_parts_from_path` helper; `to_maxtext.py` also handles `nn.LogicallyPartitioned` shape access correctly.

## Tests

- `tests/unit/tiling_test.py::LossAndGradientCorrectnessTest` — pin to Linen in `setUp` (builds via `transformer_as_linen`); drop 6 stale `pytest.skip("vocab tiling on NNX")` guards (now NNX-native via PR10).
- `tests/integration/maxengine_test.py` — drop Linen-vs-NNX prefill/decode parity tests; NNX-only assertions kept.
- `tests/unit/max_utils_test.py` — pin `UnscanTest` to Linen via `init_pyconfig`; drop the three `hasattr(state, "model")` branches (addresses [#3526 review comment](https://github.com/AI-Hypercomputer/maxtext/pull/3526#discussion_r3280576166)).
- `tests/integration/diloco_test.py` — NNX training-loop simulation + checkpoint coverage.
- `tests/integration/generate_param_only_checkpoint_test.py` — NNX param-only restore coverage.
- `tests/unit/{muon_utils,maxtext_utils,optimizers,state_dtypes,train_state_nnx_checkpoint}_test.py` — adjusted for `TrainStateNNX` / flat `nnx.State` shapes.
- `tests/unit/qwen3_next_vs_reference_test.py` — `eps` → `epsilon` rename in `Qwen3NextRMSNorm_PT` + related cleanup.
- `tests/integration/tokamax_test.py` — split parameterized test into `gmm_bf16` and `gmm_fp8` cases.

## Stats

- **Diff:** +591 / −445 across 27 files.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/layers/nnx_wrappers.py [hunk 1]
```diff
@@ -167,6 +168,31 @@ def current_linen_module() -> linen.Module | None:
...
+  Used by NNX pipeline modules to short-circuit the scan during init,
+  where only the output shape/dtype is needed.
+  """
+  module = current_linen_module()
+  if module is not None and hasattr(module, "is_initializing") and callable(module.is_initializing):
+    return module.is_initializing()
+  return False
+
+
+def _refresh_variable_trace_state(module: Module) -> None:
```
Comment (RexBearIU) [inline src/maxtext/layers/nnx_wrappers.py]: This function accesses protected members of `flax.nnx.Variable` (like `_can_update` and `_trace_state`) to manually reset trace states. This seems like a workaround for bridge limitations when using `ToNNX` with JAX tracers.
  To align with the goal of making pure NNX work as Linen without hacks, could we avoid accessing private state here? Would it be possible to ensure that `nnx.update` receives pure values without stale tracers, or work with the 
  Flax team to resolve this tracer leak issue natively in NNX? Ultimately, prioritizing the full migration of the remaining Linen modules to NNX would remove the need for this bridge and its associated workarounds.

File: src/maxtext/layers/nnx_decoders.py [hunk 2]
```diff
@@ -667,7 +673,22 @@ def layer_fn(carry, scanned_vars):
...
+        per_layer_states = []
+        for i in range(length):
+          current_params = jax.tree.map(lambda x, i=i: x[i], params)
+          current_state = jax.tree.map(lambda x, i=i: x[i], state)
+          carry, new_state_i = layer_fn(carry, (current_params, current_state))
+          per_layer_states.append(new_state_i)
+        final_carry = carry
+        scanned_state = jax.tree.map(lambda *xs: jnp.stack(list(xs)), *per_layer_states)
+      else:
+        final_carry, scanned_state = jax.lax.scan(layer_fn_wrapped, x_in, (params, state))
```
Comment (RexBearIU) [inline src/maxtext/layers/nnx_decoders.py]: Replacing `jax.lax.scan` with a Python `for` loop for FP8 is a significant performance workaround that sacrifices XLA optimizations. The root cause seems to be that Linen FP8 ops use implicit mutable scope (for things like  
  `amax_history`), which leaks JAX tracers outside the `scan` body scope, causing `UnexpectedTracerError`. Could we continue to use `jax.lax.scan` here by implementing NNX-native FP8 operations where state like `amax_history` is
  stored as a proper `nnx.Variable` (or `nnx.Cache`)? That would allow us to use the standard NNX pattern with `jax.lax.scan` (splitting state before the scan and merging it inside the body), making state movement explicit and  
  resolving the tracer leakage without resorting to a Python loop.

File: tests/unit/max_utils_test.py [hunk 3]
```diff
@@ -167,8 +168,16 @@ def test_unscan_train_state_params(self):
...
+    if hasattr(state, "model"):
+      _, params_state, _ = nnx.split(state.model, nnx.Param, ...)
+      params_to_unscan = {"params": params_state.to_pure_dict()}
+    else:
+      params_to_unscan = jax.tree_util.tree_map(lambda x: x, state.params)
+    if hasattr(sharding, "model"):
+      _, sharding_params, _ = nnx.split(sharding.model, nnx.Param, ...)
+      sharding_to_unscan = {"params": sharding_params.to_pure_dict()}
+    else:
+      sharding_to_unscan = jax.tree_util.tree_map(lambda x: x, sharding.params)
```
Comment (RexBearIU) [inline tests/unit/max_utils_test.py:194]: Here we are branching on `hasattr(state, 'model')` and converting the NNX state to a dict matching the Linen shape just to make this existing test pass for NNX. Aligned with the principle of not making workarounds in tests  
  to force old Linen tests to pass, should we consider leaving this as a pure Linen test and adding a separate test case for NNX instead? This would avoid the need to branch and manually adapt the state shape inside a test      
  designed for Linen.

File: src/maxtext/layers/nnx_wrappers.py (hunk 1 above)
Comment (Shuang-cnt) [reply src/maxtext/layers/nnx_wrappers.py]: Here is a summary of the changes made to src/maxtext/layers/nnx_wrappers.py to resolve the tracer and state mismatch issues
1. Removed Private State Access Hack
  I completely removed the _refresh_variable_trace_state function. This function was previously used as a workaround to manually force JAX tracer states to refresh by
  directly modifying protected attributes (v._can_update and v._trace_state) of flax.nnx.Variable objects.

  2. Refactored State Updates (ToLinen.__call__)
  Instead of using nnx.update(module, new_state) which caused "stale tracer" errors (because it updates variables in-place without resetting their trace level boundaries), I
  implemented the idiomatic NNX split and merge workflow.

**Review by NuojCheng — APPROVED** [review]

File: src/maxtext/configs/types.py [hunk 4]
```diff
@@ -2789,6 +2789,15 @@ def calculate_global_batch_sizes(per_device_batch_size, expansion_factor, num_de
...
     if self.using_pipeline_parallelism:
+      if self.pure_nnx:
+        # The NNX decoder has no pipeline path yet, so the scanned-layers axis ends up
+        # sharded by 'stage' and fails with a cryptic IndivisibleError at state init.
+        # Fail fast with a clear message instead. NNX pipeline support is tracked as PR11.5.
+        raise NotImplementedError(
+            "Pipeline parallelism is not yet supported on the NNX path. Set "
+            "ici_pipeline_parallelism=1 and dcn_pipeline_parallelism=1, or use the Linen path "
+            "(pure_nnx=False enable_nnx=False)."
+        )
```
Comment (bvandermoon) [inline src/maxtext/configs/types.py]: Can we merge https://github.com/AI-Hypercomputer/maxtext/pull/2885/ and remove this?

Comment (ecnal-cienet) [reply src/maxtext/configs/types.py]: Since Pipeline PR (#2885) is still being merged on Critique, I already removed this part.

**Review by bvandermoon — APPROVED** [review]
